// react-stately/dist/private/flags/flags.js
/*
 * Copyright 2023 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ let $ed892adb18214553$var$_tableNestedRows = false;
let $ed892adb18214553$var$_shadowDOM = false;
function $ed892adb18214553$export$d9d8a0f82de49530() {
    $ed892adb18214553$var$_tableNestedRows = true;
}
function $ed892adb18214553$export$1b00cb14a96194e6() {
    return $ed892adb18214553$var$_tableNestedRows;
}
function $ed892adb18214553$export$12b151d9882e9985() {
    $ed892adb18214553$var$_shadowDOM = true;
}
function $ed892adb18214553$export$98658e8c59125e6a() {
    return $ed892adb18214553$var$_shadowDOM;
}


export {$ed892adb18214553$export$d9d8a0f82de49530 as enableTableNestedRows, $ed892adb18214553$export$1b00cb14a96194e6 as tableNestedRows, $ed892adb18214553$export$12b151d9882e9985 as enableShadowDOM, $ed892adb18214553$export$98658e8c59125e6a as shadowDOM};
